Skip initial residual evaluation if possible

hugary1995 opened the pull request:
This PR deprecates the parameter `compute_initial_residual_before_preset_bcs` in favor of `use_pre_SMO_residual`.
    
The user should set this parameter to true if they really want to use the pre-initial residual in the relative convergence check. The pre-SMO residual is defined as the norm of the residual evaluated at the beginning of the time step before any solution-modifying object is executed.
    
A legacy option is added in MooseApp to make the transition smoother. All modules except Peridynamics are using the new behavior. Apps default to using the legacy behavior.
    
close #10464

hugary1995 committed:
Skip initial residual evaluation if possible

This PR deprecates the parameter `compute_initial_residual_before_preset_bcs` in favor of
`use_pre_SMO_residual`.

The user should set this parameter to true if they really want to use the pre-initial residual in the relative convergence check.
The pre-SMO residual is defined as the norm of the residual evaluated at the beginning of the time step before any solution-modifying object is executed.

A legacy option is added in MooseApp to make the transition smoother. All modules except Peridynamics are using the new behavior.
Apps default to using the legacy behavior.

close #10464

hugary1995 commented:
```
predictors/simple.test ...................................................................... FAILED (CSVDIFF)
materials/material_dependency.dont_reinit_mat_for_aux ....................................... FAILED (EXODIFF)
transfers/general_field/shape_evaluation/displaced.same_var_type ............................ FAILED (EXODIFF)
outputs/iterative.csv ....................................................................... FAILED (CSVDIFF)
predictors/simple.skip/test_skip_old ........................................................ FAILED (CSVDIFF)
postprocessors/num_residual_eval.test .......................................... [min_cpus=2] FAILED (CSVDIFF)
dirackernels/point_caching.point_caching/moving_mesh ........................................ FAILED (EXODIFF)
predictors/simple.skip_after_failed_timestep ................................................ FAILED (CSVDIFF)
materials/stateful_prop.computing_initial_residual_test ..................................... FAILED (EXODIFF)
mortar/displaced-gap-conductance-2d-non-conforming.dynamic_mortar_gap_conductance/separate ... FAILED (ERRMSG)
mortar/displaced-gap-conductance-2d-non-conforming.dynamic_mortar_gap_conductance/together ... FAILED (ERRMSG)
mortar/displaced-gap-conductance-2d-bnd-coupling.gap-conductance-bnd-material ................ FAILED (ERRMSG)
misc/jacobian.jacobian_zero .................................................. FAILED (EXPECTED ERROR MISSING)
misc/jacobian.jacobian_negative .............................................. FAILED (EXPECTED ERROR MISSING)
misc/jacobian.jacobian_zero_2D ............................................... FAILED (EXPECTED ERROR MISSING)
mortar/displaced-gap-conductance-2d-non-conforming.dynamic_gap_jacobian/separate [FINI...] FAILED (INCORRECT JACOBIAN)
mortar/displaced-gap-conductance-2d-non-conforming.dynamic_gap_conductance_fine ... [FINISHED] FAILED (ERRMSG)
mortar/displaced-gap-conductance-2d-non-conforming.linear_exact_verification ...... [FINISHED] FAILED (ERRMSG)
```

The predictor tests fail in a good way, as they are outputting the initial residual before and after preset. So it suggests this PR is taking effect. Though I still need to add a regression test specifically for this PR when all other failures are sorted out.

Interestingly, some mortar tests are failing as well.

GiudGiud commented:
I would like the default for these parameters to be dynamic. (I have not looked at the code)
If there are preset_bcs, we need to default to doing it, and have the option to not do it
Same for constraints etc

hugary1995 commented:
Why so?

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1358471/) on 6d8613f wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/23472/site)

This comment will be updated on new commits.

GiudGiud commented:
Because in the situations where it matters to do that initial evaluation, then we should do it automatically. And not miss it because the default was to not do it

lindsayad commented:
I don't know what you mean by that sentence. If we have tests that rely on side-effects of a residual evaluation in which the solution vector was in a pre-preset-bcs, etc. state, then those are bad tests. The default is to not use the residual from before preset-bc application in the relative convergence check, so in theory dropping the residual evaluation from pre-preset-bcs should not cause any issues. 

That there *are* issues/test-failures suggests there is some other problematic behavior we need to investigate.

GiudGiud commented:
I m not thinking about tests. I'm thinking about real simulations that would think they are converged if we don't perform a residual evaluation in those corner cases. 
But maybe I m mistaken, this isnt the most clear subject to me

hugary1995 commented:
Maybe think about it this way: a residual is a residual, the definition of the residual is NOT going to (at least it should not) change before/after preset BC.

So false positive convergence should be less of a concern here, at least in theory.

However, on the other hand, evaluating the initial residual pre-preset BC will indeed affect the relative convergence check, and that will potentially change the iteration behavior. I believe that's what I (we) should look into.

hugary1995 commented:
So I went ahead and fixed a bunch of the failed tests. I'll explain what I did below. Please judge whether these changes make sense:

1. The predictor tests:

```
predictors/simple.test
predictors/simple.skip/test_skip
predictors/simple.skip/test_skip_old
predictors/simple.skip_after_failed_timestep
```
These tests fail because one of the postprocessor was outputting the initial residual, and now since we are skipping the initial residual calculation whenever possible, in this case when `Predictor::shouldApply()` returns false, the postprocessor value will differ. I regolded the csv files.

2. `transfers/general_field/shape_evaluation/displaced.same_var_type`

This test failed because the subapp tries to displace the mesh based on an AuxKernel. The default execute_on for the AuxKernel is `LINEAR TIMESTEP_END`. Now since we are skipping the initial residual evaluation, the objects that are scheduled to be executed on `LINEAR` are no longer executed. I changed the execute_on of the AuxKernels to `TIMESTEP_BEGIN`, and the tests pass now -- without regolding.

3. `postprocessors/num_residual_eval.test`

This guy checks the number of residual evaluations. Indeed, the `num_residual_eval` drops from 4 to 3. So I regolded the csv file.

4. `outputs/iterative.csv`

This is a somewhat bad test. The output depends on the number of residual evaluations. I regolded it.

5. `materials/material_dependency.dont_reinit_mat_for_aux`

Another bad test. Let me directly quote the comment from the test input file:

> This test counts the number of residual evaluations that
> may slightly change from a PETSc version to another.
> For instance, starts from PETSc-3.8.4, the number of
> residual evaluating is reduced by one in a linear solver
> for each Newton iteration. This change causes this test
> fail. It  better to restrict the test
> count the residual evaluations in the nonlinear level only.

6. `materials/stateful_prop.computing_initial_residual_test`

This test uses a test material object which deliberately computes a property during initial residual evaluation. I added the parameter `compute_initial_residual_before_preset_bcs = true` to the input file, and the test passes now.

7. `dirackernels/point_caching.point_caching/moving_mesh`

This test relied on the side-effect of evaluating the initial residual. The point locator (on the displaced mesh) caches the point locations during initial residual evaluation. The fix is to displace the mesh at `TIMESTEP_BEGIN`. No need to regold the test.

With all that, the remaining failing tests are
```
mortar/displaced-gap-conductance-2d-non-conforming.dynamic_mortar_gap_conductance/separate ... FAILED (ERRMSG)
mortar/displaced-gap-conductance-2d-non-conforming.dynamic_mortar_gap_conductance/together ... FAILED (ERRMSG)
mortar/displaced-gap-conductance-2d-bnd-coupling.gap-conductance-bnd-material ................ FAILED (ERRMSG)
misc/jacobian.jacobian_zero .................................................. FAILED (EXPECTED ERROR MISSING)
misc/jacobian.jacobian_negative .............................................. FAILED (EXPECTED ERROR MISSING)
misc/jacobian.jacobian_zero_2D ............................................... FAILED (EXPECTED ERROR MISSING)
mortar/displaced-gap-conductance-2d-non-conforming.dynamic_gap_jacobian/separate . FAILED (INCORRECT JACOBIAN)
mortar/displaced-gap-conductance-2d-non-conforming.dynamic_gap_conductance_fine ... [FINISHED] FAILED (ERRMSG)
mortar/displaced-gap-conductance-2d-non-conforming.linear_exact_verification ...... [FINISHED] FAILED (ERRMSG)
```

hugary1995 commented:
8. The misc tests:
```
misc/jacobian.jacobian_zero
misc/jacobian.jacobian_negative
misc/jacobian.jacobian_zero_2D
```
fail because the test tries to displace the mesh using an AuxKernel, so they failed due to the same reason as 2. This can be easily fixed by change the execute_on of the AuxKernel.

hugary1995 commented:
9. The rest of the mortar tests fail for the same reason -- the AuxKernels displacing the mesh should be executed at `TIMESTEP_BEGIN`.

hugary1995 commented:
Some module tests fail for pretty much the same reasons.

```
combined/test:ad_cavity_pressure.values/3d .................................................. FAILED (EXODIFF)
combined/test:cavity_pressure.values/3d ..................................................... FAILED (EXODIFF)
peridynamics/test:generalized_plane_strain.gps_OSPD ......................................... FAILED (EXODIFF)
combined/test:ad_cavity_pressure.values/initial_temperature ................................. FAILED (EXODIFF)
combined/test:cavity_pressure.values/initial_temperature .................................... FAILED (EXODIFF)
peridynamics/test:generalized_plane_strain.gps_squares_OSPD ................................. FAILED (EXODIFF)
combined/test:cavity_pressure.values/rz ..................................................... FAILED (EXODIFF)
combined/test:cavity_pressure.values/multiple_postprocessors ................................ FAILED (EXODIFF)
combined/test:cavity_pressure.values/additional_volume ...................................... FAILED (EXODIFF)
combined/test:ad_cavity_pressure.values/multiple_postprocessors ............................. FAILED (EXODIFF)
combined/test:ad_cavity_pressure.values/additional_volume ................................... FAILED (EXODIFF)
fsi/test:2d-finite-strain-steady.test ....................................................... FAILED (TIMEOUT)
thermal_hydraulics/python:peacock/fluid_property_interrogator.fluid_property_interrogator ..... FAILED (CRASH)
heat_conduction/test:heat_conduction/3d_quadrature_gap_heat_transfer.moving .................. FAILED (ERRMSG)
contact/test:verification/patch_tests/ring_2.frictionless_aug ................................ FAILED (ERRMSG)
contact/test:verification/patch_tests/ring_1.frictionless_aug ................................ FAILED (ERRMSG)
contact/test:verification/patch_tests/ring_3.frictionless_aug ................................ FAILED (ERRMSG)
contact/test:verification/patch_tests/ring_4.frictionless_aug ................................ FAILED (ERRMSG)
contact/test:verification/patch_tests/cyl_3.frictionless_aug ................................. FAILED (ERRMSG)
heat_conduction/test:heat_conduction/2d_quadrature_gap_heat_transfer.moving .................. FAILED (ERRMSG)
contact/test:verification/patch_tests/ring_1.mu_0_2_aug ...................................... FAILED (ERRMSG)
contact/test:verification/patch_tests/ring_2.mu_0_2_aug ...................................... FAILED (ERRMSG)
contact/test:verification/patch_tests/ring_4.mu_0_2_aug ...................................... FAILED (ERRMSG)
```

Hopefully I can find some time this week to fix them. Feel free to review this PR meanwhile.

lindsayad commented:
> I m not thinking about tests. I'm thinking about real simulations that would think they are converged if we don't perform a residual evaluation in those corner cases.

So the change in behavior here should only make convergence more strict. The dofs that are preset/constrained will have zero residuals now. Only the non-preset dofs will be non-zero and so the initial residual norm will only include their contribution. Consequently you will be demanding that the subset of non-constrained dofs drop by the supplied relative tolerance, as opposed to the entire set. (The constrained dofs will always be converged because we are directly setting them)

GiudGiud commented on framework/src/executioners/FEProblemSolve.C:
```diff
@@ -206,6 +216,10 @@ FEProblemSolve::FEProblemSolve(Executioner & ex)
 
   _nl._compute_initial_residual_before_preset_bcs =
       getParam<bool>("compute_initial_residual_before_preset_bcs");
+  _nl._compute_initial_residual_before_constraints =
+      getParam<bool>("compute_initial_residual_before_constraints");
+  _nl._compute_initial_residual_before_predictor =
+      getParam<bool>("compute_initial_residual_before_predictor");
```
they're all different concepts. I m not sure they should be the same for the user.

BUT in practice, as soon as one is true, cant more than one become true? eg if we do before constraints, arent we doing before preset_bcs too ?

lindsayad replied:
of course they're different concepts. But as you point out if they say they want their initial residual evaluated before preset boundary conditions but not before constraints, then what should MOOSE do? Either additional logic needs to be added in this PR or we need to tell the user we can't do what they're asking for

hugary1995 replied:
Alternatively, we can consolidate these 3 parameters into 1: "compute_initial_residual".

lindsayad replied:
I would make it match the data member you renamed. `compute_initial_residual` is not as clear as your data member name

hugary1995 replied:
Oh I was thinking about also consolidating the data members.

lindsayad replied:
I mean a name like `compute_initial_residual_before_solution_modifying_objects`

hugary1995 commented:
Also I think we should activate all regression tests on this PR as this is a fairly fundamental change. I don't have the privilege though.

hugary1995 commented:
The `heat_conduction/test:interface_heating_mortar.steady_state/constraint_joule_heating_single_material` failure I don't understand. The reference residuals for the lagrange multipliers are always zero for whatever reason, so the problem never converges.

lindsayad commented:
> The modules/ad_cavity_pressure tests genuinely depend on the side effects of an initial residual evaluation. Someone will have to take a look at the regolding to make sure the results are correct.

@tophmatthews this looks like your test. Would you be willing to pull @hugary1995's branch and take a look at the re-gold?

lindsayad commented:
> The `heat_conduction/test:interface_heating_mortar.steady_state/constraint_joule_heating_single_material` failure I don't understand.

Where are you seeing this? I didn't see it on the few CIVET recipes I glanced at. That test passes for me locally and the convergence history looks identical to `devel`

lindsayad commented:
Ok I've added some more recipes

tophmatthews commented:
> @tophmatthews this looks like your test. Would you be willing to pull @hugary1995's branch and take a look at the re-gold?

I only adified the test from `combined/cavity_pressure`. The only diff (other than AD, nonAD) is `preset = false` in the AD version. Also, the diffs are pretty big, so there is genuinely something going on there that is breaking that test.

hugary1995 commented:
I think I'll just change all of the rest failing tests to use the legacy behavior.

hugary1995 commented:
I haven't forgotten this PR yet...

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1622280/) on 4ee01f1 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/23472/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/23472/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d2489239f0f53384d7fa76101f156031e46e7c5a`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1622280/) on 4ee01f1 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/23472/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/23472/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d2489239f0f53384d7fa76101f156031e46e7c5a`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1622280/) on 4ee01f1 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/23472/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/23472/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d2489239f0f53384d7fa76101f156031e46e7c5a`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1622280/) on 4ee01f1 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/23472/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/23472/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d2489239f0f53384d7fa76101f156031e46e7c5a`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1622280/) on 4ee01f1 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/23472/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/23472/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d2489239f0f53384d7fa76101f156031e46e7c5a`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1622280/) on 4ee01f1 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/23472/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/23472/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d2489239f0f53384d7fa76101f156031e46e7c5a`

hugary1995 commented on framework/src/predictors/SimplePredictor.C:
```diff
@@ -32,9 +32,6 @@ SimplePredictor::shouldApply()
   if (_t_step < 2 || _dt_old <= 0)
     should_apply = false;
 
-  if (!should_apply)
-    _console << "  Skipping predictor this step" << std::endl;
```
I think in general we should allow this test method `shouldApply()` to be called multiple times at various places. Certainly we don't want to see this console message _every time_ this method is called. I think it is the caller's responsible to print this message.

I have moved this console message to `NonlinearSystemBase::setInitialSolution()`.

hugary1995 commented:
Sorry for the delay. I am finally coming back to this.

@lindsayad I've addressed your comments and also made a few more changes/fixes/renamings. Maybe first read the updated pull request description to refresh your memory.

I've fixed all the framework and modules tests that fail on my local machine. I've tried my best to make the "correct" changes and regolded as needed. 

One exception is in the (ad)cavity_pressure tests in the combined module, which as @tophmatthews suggested there's something wonky going on. I have set `use_pre_initial_residual = true` in those tests so that it falls back to the legacy behavior (evaluating the pre-preset BC residual for the convergence check).

hugary1995 commented:
Okay, I looked into the cavity pressure tests. The issue comes from the CavityPressureUOAction which sets the execute_on of the CavityPressureUserObject to be linear, and the CavityPressureUserObject has a weirdo `_initialized` flag which makes the simulation results depend on the side effects of pre-preset-bc residual evaluation.

I don't see a nice way of fixing that without regolding, and I don't want to regold without the permission of the original creater of these CavityPressure classes and tests.

hugary1995 commented:
Two of the solution invalidity tests are generating different jsons in serial and in parallel, which I have no clue on...
```
misc/solution_invalid.solution_invalid_transient ........................................... FAILED (JSONDIFF)
misc/solution_invalid.test/solution_valid_recover .......................................... FAILED (JSONDIFF)
```

lindsayad commented:
@MengnanLi91 could you potentially help @hugary1995 with the solution invalidity failures?

MengnanLi91 commented:
> @MengnanLi91 could you potentially help @hugary1995 with the solution invalidity failures?

Sure, I'll take a look. Gary and I have scheduled a meeting tomorrow morning.

hugary1995 commented:
> @MengnanLi91 could you potentially help @hugary1995 with the solution invalidity failures?

Yeah, we're having a talk tomorrow morning.

hugary1995 commented:
My last commit should fix this. @MengnanLi91 please review the changes I made when you get a chance.

MengnanLi91 commented on test/tests/misc/solution_invalid/gold/solution_invalid_recover.json:
The timestep counts doesn't seem right. In the test case, for each iteration, the count is 16. Each time step has 2 linear and 2 non-linear iterations, I think the total count for each time step should be 64 instead of 48. Do I miss something here?

hugary1995 replied:
There are two residual evaluations and one Jacobian evaluation. So (2+1)*16=48

hugary1995 replied:
@lindsayad see my comment above -- this is me whining again about the namings :J

MengnanLi91 replied:
Okay, I thought there are two nonlinear/Jacobian evaluations. Thanks for the explanation! The results make sense to me now.

lindsayad replied:
Oh I don't like the namings either. I didn't pick them. That was before my time 😄

lindsayad commented:
@hugary1995 are you able to investigate the bison failure?

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1624735/) on 61cb5c4 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>d24892</th>
      <th colspan=4>#23472 61cb5c</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.60%</td>
      <td>85.57%</td>
      <td>-0.03%</td>
      <td>81.36%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>89097</td>
      <td>89088</td>
      <td>-9</td>
      <td>48</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14994</td>
      <td>15023</td>
      <td>+29</td>
      <td>11</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23472/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23472/coverage/framework)


## Modules coverage

### Tensor mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>d24892</th>
      <th colspan=4>#23472 61cb5c</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.66%</td>
      <td>84.65%</td>
      <td>-0.01%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>25443</td>
      <td>25439</td>
      <td>-4</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>4610</td>
      <td>4614</td>
      <td>+4</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23472/coverage_diff/tensor_mechanics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23472/coverage/tensor_mechanics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/23472/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/23472/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/23472/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/23472/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/23472/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/23472/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/23472/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/23472/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/23472/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/23472/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/23472/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/23472/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/23472/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/23472/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/23472/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/23472/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/23472/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/23472/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/23472/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/23472/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/23472/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/23472/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/23472/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/23472/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/23472/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/23472/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/23472/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/23472/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 81.36% is less than the suggested 90.0%



This comment will be updated on new commits.

lindsayad commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -220,24 +220,23 @@ class FEProblemBase : public SubProblem, public Restartable
    * @param max_funcs      Maximum Number of function evaluations
-   * @param initial_residual_before_preset_bcs      Residual norm prior to imposition of preset BCs
-   * values on solution vector
+   * @param pre_initial_residual      Residual norm prior to imposition of solution modifying
```
This spacing is weird. I would just do a single space between all the params and their documentation strings

hugary1995 commented:
> @hugary1995 are you able to investigate the bison failure?

No I can't access.

lindsayad commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -220,24 +220,23 @@ class FEProblemBase : public SubProblem, public Restartable
+                                                                    const Real abstol,
+                                                                    const PetscInt nfuncs,
+                                                                    const PetscInt max_funcs,
+                                                                    const Real pre_initial_residual,
```
It's long but I might name this something like `residual_before_solution_modifying_objects`. `pre_initial_residual` is a bit odd to me because it is a residual

lindsayad commented on framework/include/systems/NonlinearSystemBase.h:
```diff
@@ -556,6 +556,21 @@ class NonlinearSystemBase : public SystemBase, public PerfGraphInterface
+   *
+   * \return A boolean indicating whether we should evaluate the pre-initial residual.
+   */
+  bool shouldEvaluatePreinitialResidual() const;
```
```suggestion
  bool shouldEvaluateResidualBeforeSolutionModifyingObjects() const;
```
how about that for a mouthful! Again I think something like this is more clear as I would argue that if you do evaluate this residual, then it is chronologically the initial residual

lindsayad commented on framework/include/systems/NonlinearSystemBase.h:
```diff
@@ -682,11 +697,14 @@ class NonlinearSystemBase : public SystemBase, public PerfGraphInterface
   Real _last_nl_rnorm;
-  Real _initial_residual_before_preset_bcs;
-  Real _initial_residual_after_preset_bcs;
+  /// The initial residual _before_ executing solution modifying objects
+  Real _pre_initial_residual;
+  /// The initial residual _after_ executing solution modifying objects
+  Real _initial_residual;
```
```suggestion
  /// The initial residual _before_ executing solution modifying objects
  Real _initial_residual_before_solution_modifying_objects;
  /// The initial residual _after_ executing solution modifying objects
  Real _initial_residual_after_solution_modifying_objects;
```

lindsayad commented on framework/include/systems/NonlinearSystemBase.h:
```diff
@@ -682,11 +697,14 @@ class NonlinearSystemBase : public SystemBase, public PerfGraphInterface
   unsigned int _current_nl_its;
-  bool _compute_initial_residual_before_preset_bcs;
+  /// Whether to use the pre-initial residual in the relative convergence check
+  bool _use_pre_initial_residual;
```
```suggestion
  bool _use_initial_residual_before_solution_modifying_objects;
```

lindsayad reviewed (commented):
I'm offering my opinion on the member names

hugary1995 commented:
I don't have strong opinion on the member names, so I guess I'll just go with your suggestions. I'll also try to make a forward compatible patch in the external apps over the holidays. Someone else will have to take a look at the internal apps.

hugary1995 commented:
I ended up renaming "pre_initial_residual" to "fnorm0_before_smo" and "initial_residual" to "fnorm0_after_smo".

I choose "fnorm" as that is consistent with other references to the initial residual in FEProblemBase and NonlinearSystemBase. "solution_modifying_objects" seems a bit too long, so I used "smo" instead (we also have various other acronyms in member names in the same files so...).

lindsayad commented:
K, I like those names better than before 👍

hugary1995 commented:
I'll come back to this eventually.

hugary1995 commented:
Thanks for the reminder lol. At this point, I think I just add an option to moose app which defaults to the legacy behavior. I'll change the modules that pass the tests to the new behavior.

hugary1995 commented:
I swear I'll never procrastinate writing pull request in the future. I have to review my own pull request to remember what I did.

loganharbour commented:
> I swear I'll never procrastinate writing pull request in the future. I have to review my own pull request to remember what I did.

![hehehe](https://github.com/idaholab/moose/assets/22062133/91ea6f6d-6401-41d9-81fd-233bc3dbb38e)

hugary1995 commented:
lol it automatically triggered review requests for everyone because I modified all `*App.C` in the repo. Feel free to comment or just ignore the request.

hugary1995 commented:
@lindsayad, this is ready for review.

loganharbour commented:
Have you seen what the effect is to make this the default? If not, we should have another PR and change the default just to look at the testing.

hugary1995 commented:
Yes, I have. Right now everything in the moose repo should be using the new behavior. All other apps default to legacy behavior.

Most of the apps can pass with the new behavior, except a few such as Zapdos, MARMOT, etc. But only a handful of tests diff. Fixing them would require a bit domain expertise.

loganharbour commented:
I think it might be worth starting that conversation now and tabling apps that can be done easy (those that pass) and those that can't

hugary1995 commented:
I am not sure if the "Controlled app tests" failure is related to this PR.

hugary1995 commented:
Okay, so according to #26222

Public apps that will need to be fixed
- Mastodon: easy fix
- Zapdos: will need someone with domain knowledge to investigate
- Blackbear: easy fix

At least one internal app will need to be fixed, but I can see the test results.

lindsayad commented:
> * Zapdos: will need someone with domain knowledge to investigate

I can do that if you create an issue on the repo

hugary1995 commented:
Here you go: https://github.com/shannon-lab/zapdos/issues/217

lindsayad commented on framework/src/systems/NonlinearSystem.C:
```diff
@@ -212,15 +213,15 @@ NonlinearSystem::solve()
   // output the solution invalid summary
   if (_solution_is_invalid)
   {
-    // sync all solution invalid counts to rank 0 process
-    _app.solutionInvalidity().sync();
```
This seems important?

lindsayad replied:
Never mind I see the new lines in the base class

hugary1995 commented on framework/src/systems/NonlinearSystemBase.C:
```diff
@@ -3589,6 +3596,56 @@ NonlinearSystemBase::setMooseKSPNormType(MooseEnum kspnorm)
 }
 
+bool
+NonlinearSystemBase::shouldEvaluateFNormBeforeSMO() const
+{
+  // This complies with the legacy behavior: always do the redundant residual evaluation
+  if (_app.parameters().get<bool>("use_legacy_initial_residual_evaluation_bahavior"))
+    return true;
+
+  // If solve_type == LINEAR, there is no need to evaluate the initial residual.
+  if (_fe_problem.solverParams()._type == Moose::ST_LINEAR)
+    return false;
+
+  // If the user explicitly says no to "computing pre-initial residual", then directly return false
+  if (!_use_fnorm0_before_smo)
+    return false;
+
+  // Otherwise, if there are active solution-modifying objects, we compute the pre-initial residual
+  // and use it in the relative convergence check.
+  if (hasActiveSolutionModifyingObjects())
+    return true;
```
- correct
- and correct

> (because why would we ... the result should be the same)

That's the hope. But if we are not careful enough in accounting for all the SMOs, the solution may change. For example, what if there is a userobject that modifies the solution vector? And in that case, we will have to rely on the user to set `use_fnorm_before_smo` to the value they want.

hugary1995 replied:
Okay, so I've modified the logic a little bit (see [87526cd](https://github.com/idaholab/moose/pull/23472/commits/87526cd76974cc6c8df3eec830ecb6a90302d10d)) -- if the user explicitly sets this parameter, we will respect the user's choice regardless of the existence of SMOs.

lindsayad replied:
ok that's good because it seems like it would have been ignored before if there weren't the few SMOs we're aware of at the framework level...?

hugary1995 replied:
That's right.

hugary1995 commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -7786,11 +7786,13 @@ FEProblemBase::checkNonlinearConvergence(std::string & msg,
-                            : system._initial_residual_after_preset_bcs;
+    // Set the initial residual depending on what the user asks us to use.
+    Real the_residual;
+    if (_app.parameters().get<bool>("use_legacy_initial_residual_evaluation_bahavior"))
+      the_residual = system._use_fnorm0_before_smo ? fnorm0_before_smo : system._fnorm0_after_smo;
+    else
+      the_residual =
+          system.shouldEvaluateFNormBeforeSMO() ? fnorm0_before_smo : system._fnorm0_after_smo;
```
I am not sure if it is possible. The main issue is that the legacy behavior is not self consistent: it chooses which residual to use according to this input parameter, but it always performs the redundant residual evaluation (even if it will not be used).

hugary1995 commented on framework/src/executioners/FEProblemSolve.C:
```diff
@@ -87,11 +87,20 @@ FEProblemSolve::validParams()
   params.addParam<bool>(
       "skip_exception_check", false, "Specifies whether or not to skip exception check");
-  params.addParam<bool>(
+  params.addDeprecatedParam<bool>(
```
yeah that's a better way of handling these parameters.

hugary1995 commented:
@lindsayad I'm not in a hurry, but I think the ball is in your court: https://github.com/shannon-lab/zapdos/issues/217

Also: go away stale bot.

lindsayad commented:
@hugary1995 I just checked out shannon-lab/zapdos@8ca2c716 which was the last commit pulled by the test harness on this PR and after fixing the solver group parameters in `FEProblemSolve.C` I don't see any diffs in Zapdos

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2156055/) on 98f94d5 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/23472/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/23472/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b02bb194f5ca6361a22e45948575b03c3d10f698`

lindsayad commented:
If you have the defaulting to legacy, then I'm all for full steam ahead on this PR. The issue on Zapdos can remain open

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2156520/) on e84dc3a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/23472/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/23472/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b02bb194f5ca6361a22e45948575b03c3d10f698`

hugary1995 commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -215,35 +215,31 @@ class FEProblemBase : public SubProblem, public Restartable
+   * @param max_funcs Maximum Number of function evaluations
+   * @param div_threshold Maximum value of residual before triggering divergence check
+   */
+  virtual MooseNonlinearConvergenceReason checkNonlinearConvergence(std::string & msg,
+                                                                    const PetscInt it,
+                                                                    const Real xnorm,
+                                                                    const Real snorm,
+                                                                    const Real fnorm,
+                                                                    const Real rtol,
+                                                                    const Real divtol,
+                                                                    const Real stol,
+                                                                    const Real abstol,
+                                                                    const PetscInt nfuncs,
+                                                                    const PetscInt max_funcs,
+                                                                    const Real div_threshold);
```
It is no longer needed to pass `initial_residual_before_preset_bcs` here, because the `NonlinearSystemBase` knows the reference residual to use. See my changes in `NonlinearSystemBase.h`.

hugary1995 commented on framework/include/systems/NonlinearSystemBase.h:
```diff
@@ -227,6 +227,45 @@ class NonlinearSystemBase : public SolverSystem, public PerfGraphInterface
    */
   std::shared_ptr<Split> getSplit(const std::string & name);
 
+  /**
+   * We offer the option to check convergence against the pre-SMO residual. This method handles the
+   * logic as to whether we should perform such residual evaluation.
+   *
+   * @return A boolean indicating whether we should evaluate the pre-SMO residual.
+   */
+  bool shouldEvaluatePreSMOResidual() const;
+
+  /**
+   * Returns a writable reference to a boolean controlling whether to evaluate the pre-SMO residual
+   * and use it in the subsequent relative convergence checks.
+   *
+   * If set to true, an _additional_ residual evaluation is performed before any
+   * solution-modifying object is executed, and before the initial (0-th nonlinear iteration)
+   * residual evaluation. Such residual is referred to as the pre-SMO residual. If the pre-SMO
+   * residual is evaluted, it is used in the subsequent relative convergence checks.
+   *
+   * If set to false, no residual evaluation takes place before the initial residual evaluation, and
+   * the initial residual is used in the subsequent relative convergence checks. This mode is
+   * recommended for performance-critical code as it avoids the additional pre-SMO residual
+   * evaluation.
+   */
+  bool & usePreSMOResidual() { return _use_pre_smo_residual; }
```
In principle we don't need both methods, because `shouldEvaluatePreSMOResidual()` and `usePreSMOResidual()` should always return the same value. However, we need both for now to stay consistent with the legacy behavior (where an additional residual evaluation always takes place).

hugary1995 commented on framework/include/systems/NonlinearSystemBase.h:
```diff
@@ -873,6 +909,13 @@ class NonlinearSystemBase : public SolverSystem, public PerfGraphInterface
 
   bool _computing_initial_residual;
 
+  /// The pre-SMO residual
+  Real _pre_smo_residual;
+  /// The initial residual
+  Real _initial_residual;
```
I named them this way so that there is no conceptual clash between the two residuals. "Initial residual is initial residual".

lindsayad replied:
Can you clarify in the doxygen whether this may be the pre-SMO residual or the post-SMO residual?

hugary1995 commented on framework/src/executioners/EigenExecutionerBase.C:
```diff
@@ -263,7 +263,7 @@ EigenExecutionerBase::inversePowerIteration(unsigned int min_iter,
     // save the initial residual
     if (iter == 0)
-      initial_res = _eigen_sys._initial_residual_before_preset_bcs;
+      initial_res = _eigen_sys.referenceResidual();
```
This is a real change, and I think this is a right change.

This could potentially lead to some diffs in theory, but I am not seeing any.

hugary1995 committed:
Remove unnecessary includes

hugary1995 committed:
Apply format change

hugary1995 committed:
Fix framework tests

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2156665/) on f76f433 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/23472/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/23472/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b02bb194f5ca6361a22e45948575b03c3d10f698`

hugary1995 committed:
Annoying clang-format

hugary1995 committed:
Peridynamics should use legacy behavior

hugary1995 committed:
peridynamics use new behavior, tighten test tolerances

hugary1995 committed:
peridynamics does not work with automatic scaling

hugary1995 committed:
Fix BatchMaterial

hugary1995 committed:
Uncovered a bug with postResidual time resetting

hugary1995 committed:
Revert "Uncovered a bug with postResidual time resetting"

This reverts commit df0b6b0b8b0297f8ff4d39e59db4b67e3bfe67d0.

hugary1995 committed:
Skip pre-SMO residual evaluation if solve type is linear

hugary1995 commented:
@lindsayad This is ready for another review.

lindsayad commented on framework/include/systems/NonlinearSystemBase.h:
```diff
@@ -86,10 +86,10 @@ class NonlinearSystemBase : public SolverSystem, public PerfGraphInterface
+   * @return Whether or not we are currently computing the pre-SMO residual.
    */
-  virtual bool computingInitialResidual() { return _computing_initial_residual; }
+  virtual bool computingPreSMOResidual() { return _computing_pre_smo_residual; }
```
does anyone override this? If not, let's remove the virtual

lindsayad commented on framework/include/systems/NonlinearSystemBase.h:
```diff
@@ -227,6 +227,45 @@ class NonlinearSystemBase : public SolverSystem, public PerfGraphInterface
+   * recommended for performance-critical code as it avoids the additional pre-SMO residual
+   * evaluation.
+   */
+  bool & usePreSMOResidual() { return _use_pre_smo_residual; }
```
can we change this to accept an argument to perform the write that way instead of releasing the data member as a writable reference? This allows some degree of encapsulation such that if we need to add some sanity checks (or other things) in the future, we can do so. We  have discovered the hard way in libMesh that by releasing writable references to data members, we have no way of tracking whether a class  has been modified or not

hugary1995 replied:
> we have no way of tracking whether a class has been modified or not

well you could return a wrapper around boolean e.g. `Ref<bool>` that overloads `operator=`, but I see your point.

lindsayad reviewed (commented):
very close

hugary1995 committed:
Address comments

hugary1995 commented:
@lindsayad I'll let you hit the merge button as I can't see the failures. Feel free to squash all the commits into the first one.

lindsayad commented:
I will wait for the Griffin MR to make it to master and then I will invalidate the targets here. Should be able to merge all green

moosebuild commented:
Job [Controlled app tests](https://civet.inl.gov/job/2162222/) on c0def29 : invalidated by @lindsayad

moosebuild commented:
Job [Documentation controlled apps](https://civet.inl.gov/job/2162191/) on c0def29 : invalidated by @lindsayad

lindsayad commented:
This is causing a valgrind failure for a navier-stokes module test
```
test:finite_element/pins/expansion-channel.slip_wall
```
Sadly, the backtrace is really not very helpful
```
==2731129== Conditional jump or move depends on uninitialised value(s)
==2731129==    at 0x16482996: MatSetValues_SeqAIJ (aij.c:429)
==2731129==    by 0x166C31FB: MatSetValues (matrix.c:1497)
==2731129==    by 0x1475C5C1: libMesh::PetscMatrix<double>::add(unsigned long, unsigned long, double) (petsc_matrix.C:1244)
==2731129==    by 0x9DFE027: Assembly::addCachedJacobian(Assembly::GlobalDataKey) (Assembly.C:3825)
==2731129==    by 0xB642553: SubProblem::addCachedJacobian(unsigned int) (SubProblem.C:1285)
==2731129==    by 0xB5D0D8B: FEProblemBase::addCachedJacobian(unsigned int) (FEProblemBase.C:1790)
==2731129==    by 0xB27183C: ComputeJacobianThread::postElement(libMesh::Elem const*) (ComputeJacobianThread.C:155)
==2731129==    by 0x9EC8215: ThreadedElementLoopBase<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> >::operator()(libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> const&, bool) (ThreadedElementLoopBase.h:295)
==2731129==    by 0xB282286: NonlinearThread::operator()(libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> const&, bool) (NonlinearThread.C:60)
==2731129==    by 0xBE83A5E: void libMesh::Threads::parallel_reduce<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*>, ComputeFullJacobianThread>(libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> const&, ComputeFullJacobianThread&) (threads_pthread.h:370)
==2731129==    by 0xBE4C427: NonlinearSystemBase::computeJacobianInternal(std::set<unsigned int, std::less<unsigned int>, std::allocator<unsigned int> > const&) (NonlinearSystemBase.C:2869)
==2731129==    by 0xBE4E1B6: NonlinearSystemBase::computeJacobianTags(std::set<unsigned int, std::less<unsigned int>, std::allocator<unsigned int> > const&) (NonlinearSystemBase.C:3093)
==2731129==    by 0xB61A7F3: FEProblemBase::computeJacobianTags(std::set<unsigned int, std::less<unsigned int>, std::allocator<unsigned int> > const&) (FEProblemBase.C:6854)
==2731129==    by 0xB6196FB: FEProblemBase::computeJacobianInternal(libMesh::NumericVector<double> const&, libMesh::SparseMatrix<double>&, std::set<unsigned int, std::less<unsigned int>, std::allocator<unsigned int> > const&) (FEProblemBase.C:6780)
==2731129==    by 0xB618ED4: FEProblemBase::computeJacobian(libMesh::NumericVector<double> const&, libMesh::SparseMatrix<double>&, unsigned int) (FEProblemBase.C:6766)
==2731129==    by 0xB618C1D: FEProblemBase::computeJacobianSys(libMesh::NonlinearImplicitSystem&, libMesh::NumericVector<double> const&, libMesh::SparseMatrix<double>&) (FEProblemBase.C:6736)
==2731129==    by 0xBE2E07C: Moose::compute_jacobian(libMesh::NumericVector<double> const&, libMesh::SparseMatrix<double>&, libMesh::NonlinearImplicitSystem&) (NonlinearSystem.C:39)
==2731129==    by 0xBE2FBE2: NonlinearSystem::setupColoringFiniteDifferencedPreconditioner() (NonlinearSystem.C:280)
==2731129==    by 0xBE2F986: NonlinearSystem::setupFiniteDifferencedPreconditioner() (NonlinearSystem.C:235)
==2731129==    by 0xBE2F294: NonlinearSystem::solve() (NonlinearSystem.C:164)
==2731129==    by 0xB609BED: FEProblemBase::solve(unsigned int) (FEProblemBase.C:5929)
==2731129==    by 0xACA5854: FEProblemSolve::solve() (FEProblemSolve.C:293)
==2731129==    by 0xACAAD72: FixedPointSolve::solveStep(double&, double&, std::set<unsigned long, std::less<unsigned long>, std::allocator<unsigned long> > const&) (FixedPointSolve.C:439)
==2731129==    by 0xACA9B99: FixedPointSolve::solve() (FixedPointSolve.C:291)
==2731129==    by 0xAD27D41: TimeStepper::step() (TimeStepper.C:168)
==2731129==    by 0xACB9C43: Transient::takeStep(double) (Transient.C:416)
==2731129==    by 0xACB86B3: Transient::execute() (Transient.C:291)
==2731129==    by 0x9E393A3: MooseApp::executeExecutioner() (MooseApp.C:1161)
==2731129==    by 0x9E4056F: MooseApp::run() (MooseApp.C:1543)
==2731129==    by 0x1169CF: void Moose::main<NavierStokesTestApp>(int, char**) (MooseMain.h:47)
==2731129==    by 0x1156DC: main (main.C:17)
==2731129==  Uninitialised value was created by a heap allocation
==2731129==    at 0x4843FA3: operator new(unsigned long) (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==2731129==    by 0xB764AC2: std::__new_allocator<std::_Sp_counted_ptr_inplace<NonlinearSystem, std::allocator<void>, (__gnu_cxx::_Lock_policy)2> >::allocate(unsigned long, void const*) (new_allocator.h:137)
==2731129==    by 0xB743440: std::allocator_traits<std::allocator<std::_Sp_counted_ptr_inplace<NonlinearSystem, std::allocator<void>, (__gnu_cxx::_Lock_policy)2> > >::allocate(std::allocator<std::_Sp_counted_ptr_inplace<NonlinearSystem, std::allocator<void>, (__gnu_cxx::_Lock_policy)2> >&, unsigned long) (alloc_traits.h:464)
==2731129==    by 0xB718FC7: std::__allocated_ptr<std::allocator<std::_Sp_counted_ptr_inplace<NonlinearSystem, std::allocator<void>, (__gnu_cxx::_Lock_policy)2> > > std::__allocate_guarded<std::allocator<std::_Sp_counted_ptr_inplace<NonlinearSystem, std::allocator<void>, (__gnu_cxx::_Lock_policy)2> > >(std::allocator<std::_Sp_counted_ptr_inplace<NonlinearSystem, std::allocator<void>, (__gnu_cxx::_Lock_policy)2> >&) (allocated_ptr.h:98)
==2731129==    by 0xB6E6A0E: std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count<NonlinearSystem, std::allocator<void>, FEProblem&, NonlinearSystemName const&>(NonlinearSystem*&, std::_Sp_alloc_shared_tag<std::allocator<void> >, FEProblem&, NonlinearSystemName const&) (shared_ptr_base.h:969)
==2731129==    by 0xB6B12B7: std::__shared_ptr<NonlinearSystem, (__gnu_cxx::_Lock_policy)2>::__shared_ptr<std::allocator<void>, FEProblem&, NonlinearSystemName const&>(std::_Sp_alloc_shared_tag<std::allocator<void> >, FEProblem&, NonlinearSystemName const&) (shared_ptr_base.h:1712)
==2731129==    by 0xB67D332: std::shared_ptr<NonlinearSystem>::shared_ptr<std::allocator<void>, FEProblem&, NonlinearSystemName const&>(std::_Sp_alloc_shared_tag<std::allocator<void> >, FEProblem&, NonlinearSystemName const&) (shared_ptr.h:464)
==2731129==    by 0xB65209D: std::shared_ptr<std::enable_if<!std::is_array<NonlinearSystem>::value, NonlinearSystem>::type> std::make_shared<NonlinearSystem, FEProblem&, NonlinearSystemName const&>(FEProblem&, NonlinearSystemName const&) (shared_ptr.h:1010)
==2731129==    by 0xB5B38B8: FEProblem::FEProblem(InputParameters const&) (FEProblem.C:41)
==2731129==    by 0xB7B2466: std::__detail::_MakeUniq<FEProblem>::__single_object std::make_unique<FEProblem, InputParameters const&>(InputParameters const&) (unique_ptr.h:1065)
==2731129==    by 0xB7AA967: RegistryEntry<FEProblem>::build(InputParameters const&) (Registry.h:276)
==2731129==    by 0x9E088E8: Factory::createUnique(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, InputParameters const&, unsigned int, bool) (Factory.C:101)
==2731129==    by 0xAE354B4: std::unique_ptr<FEProblemBase, std::default_delete<FEProblemBase> > Factory::createUnique<FEProblemBase>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, InputParameters const&, unsigned int) (Factory.h:254)
==2731129==    by 0xAE220DD: std::shared_ptr<FEProblemBase> Factory::create<FEProblemBase>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, InputParameters const&, unsigned int) (Factory.h:270)
==2731129==    by 0xADCADF8: CreateProblemDefaultAction::act() (CreateProblemDefaultAction.C:82)
==2731129==    by 0xADA59A0: Action::timedAct() (Action.C:86)
==2731129==    by 0xADAA31E: ActionWarehouse::executeActionsWithAction(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (ActionWarehouse.C:388)
==2731129==    by 0xADA9E48: ActionWarehouse::executeAllActions() (ActionWarehouse.C:347)
==2731129==    by 0x9E38292: MooseApp::runInputFile() (MooseApp.C:1106)
==2731129==    by 0x9E3FDC9: MooseApp::run() (MooseApp.C:1533)
==2731129==    by 0x1169CF: void Moose::main<NavierStokesTestApp>(int, char**) (MooseMain.h:47)
==2731129==    by 0x1156DC: main (main.C:17)
```
